feat(docs): Added docs about using Metro for web (#18417)
expo/expo: docs/pages/guides/customizing-metro.md

title: Customizing Metro
---

> import { Terminal } from '~/ui/components/Snippet';

Reviewer: ```suggestion
import { Terminal } from '~/ui/components/Snippet';
import { YesIcon, NoIcon } from '~/ui/components/DocIcons';
```

To add to a value, such as an array of file extensions, defined in the default configuration, you can access the defaults using the `getDefaultConfig(projectRoot)` function defined in `@expo/metro-config`. Add `@expo/metro-config` to the dependencies of your project to do this. In SDK +41 you can import from `expo/metro-config`.

## Web Support

> Metro web support is an experimental feature in Expo SDK 46 and only works with the [local Expo CLI](https://blog.expo.dev/new-versioned-expo-cli-cf6e10632656).

> By default, Expo CLI uses Webpack as the bundler on web platforms, this is because Metro historically did not support web.
> Using different bundlers across platforms leads to some critical divergence in how your app works across platforms. Features like Fast Refresh which work on native don't work on web, and important production functionality like assets are treated differently across bundlers.

Reviewer: ```suggestion
By default, Expo CLI uses Webpack as the bundler on web platforms, this is because Metro historically did not support web. Using different bundlers across platforms leads to some critical divergence in how your app works across platforms. Features like Fast Refresh which work on native don't work on web, and important production functionality like assets are treated differently across bundlers.
```

I think it's the prettier formatting not working here. I think those two sentences mean to be in the same paragraph.

By utilizing Metro across all platforms you can have a more universal development experience. You also get to utilize shared cached chunks across platforms meaning faster iteration speed when working across platforms. Project upgrades can also be easier since there are less dependencies (`webpack`, `webpack-dev-server`) you need to update between versions.

All Metro web features should be universal (bundling, static files, assets) making the DX easier to understand and faster across the app development process.

Learn once, bundle everywhere!

### Expo Webpack vs. Expo Metro

> Universal Expo Metro is designed to be fully universal, meaning any web bundling features should also work on native too.
> Because of this we make some breaking changes between the two bundler implementations, carefully check the difference if you're moving from Webpack to Metro.
> 

Reviewer: ```suggestion
Universal Expo Metro is designed to be fully universal, meaning any web bundling features should also work on native too. Because of this we make some breaking changes between the two bundler implementations, carefully check the difference if you're moving from Webpack to Metro.

```

> | Feature          | Metro                | Webpack                |
> | ---------------- | -------------------- | ---------------------- |
> | Start command    | `npx expo start`     | `npx expo start`       |
> | Bundle command   | `npx expo export`    | `npx expo export:web`  |
> | Output folder    | `dist/`              | `web-build/`           |
> | Static folder    | `public/`            | `web/`                 |
> | Config file      | `metro.config.js`    | `webpack.config.js`    |
> | Default config   | `@expo/metro-config` | `@expo/webpack-config` |
> | Fast Refresh     | WIP (coming soon)    | ❌                     |
> | Tree Shaking     | ❌                   | ✅                     |
> | CSS Handling     | ❌                   | ✅                     |
> | Asset Manifests  | ❌                   | ✅                     |
> | Bundle Splitting | WIP (pending native) | ✅                     |

Reviewer: ```suggestion
| Feature          | Metro                | Webpack                |
| ---------------- | -------------------- | ---------------------- |
| Start command    | `npx expo start`     | `npx expo start`       |
| Bundle command   | `npx expo export`    | `npx expo export:web`  |
| Output folder    | `dist/`              | `web-build/`           |
| Static folder    | `public/`            | `web/`                 |
| Config file      | `metro.config.js`    | `webpack.config.js`    |
| Default config   | `@expo/metro-config` | `@expo/webpack-config` |
| Fast Refresh     | WIP (coming soon)    | <NoIcon />             |
| Tree Shaking     | <NoIcon />           | <YesIcon />            |
| CSS Handling     | <NoIcon />           | <YesIcon />            |
| Asset Manifests  | <NoIcon />           | <YesIcon />            |
| Bundle Splitting | WIP (pending native) | <YesIcon />            |
```

Let's use `NoIcon` and `YesIcon` components instead of emojis. Also, added the import statement for these components at the top of this page.

```

You can skip bundling for all platforms by using the `--platform` flag:

<Terminal cmd={["$ npx expo export --platform web"]} />

Your output will be found in the `dist` folder.

The output can be tested locally by using the [Serve CLI](https://www.npmjs.com/package/serve).

<Terminal cmd={["$ npx serve dist"]} />

You can deploy the website using any popular web host, follow any of the guides in [publishing websites](/distribution/publishing-websites/), just substitute the `web-build/` folder for the `dist/` folder.

#### Static Files

Expo's Metro implementation supports hosting static files from the dev server by putting them in the root `public/` folder, this is akin to many other web frameworks. In Expo Webpack we default to using the `web/` folder.

When exporting with `npx expo export` we copy the contents of the `public/` folder into the `dist/` folder, meaning your app can expect to fetch these assets relative to the host URL.

The most common example of this is the `public/favicon.ico` which is used by websites to render the tab icon.

You can overwrite the default `index.html` in Metro web by creating a `public/index.html` file in your project.

> In the future, this will work universally across platforms with EAS Update hosting. Currently the feature is web-only based on the static host used for the native app, e.g. the legacy Expo service updates does not support this feature.

Reviewer: ```suggestion
In the future, this will work universally across platforms with EAS Update hosting. Currently the feature is web-only based on the static host used for the native app, for example, the legacy Expo service updates does not support this feature.
```

Let's use the full form instead of Latin abbreviation here 👍
